// microsoft-servicehub/hubLookup.js
"use strict";
var q_1 = require('q');
var logger_1 = require('./logger');
var jsonRpc_1 = require('./jsonRpc');
var locationService_1 = require('./locationService');
var controllerConnection_1 = require('./controllerConnection');
var common_1 = require('./common');
var ec = require('errno-codes');
var HubLookup = (function () {
    function HubLookup(options) {
        this.options = options || {};
        this.options.clientName = this.options.clientName || 'hubLookup';
        this.options.logger = this.options.logger || new logger_1.Logger(this.options.clientName);
    }
    HubLookup.prototype.getLocationService = function () {
        var _this = this;
        this.throwIfDisposed();
        if (!this.locationService) {
            this.locationService = controllerConnection_1.ControllerConnection.connect(this.options)
                .then(function (connection) {
                _this.connection = connection;
                var rpc = jsonRpc_1.JsonRpcConnection.attach(connection.stream, _this.options.logger);
                rpc.onClose(_this.resetLocationService, _this);
                return new locationService_1.LocationServiceProxy(rpc);
            }, function (err) {
                _this.resetLocationService();
                throw err;
            });
        }
        return this.locationService;
    };
    HubLookup.prototype.getConfig = function () {
        return this.options.getConfig();
    };
    HubLookup.prototype.stopHubController = function () {
        var _this = this;
        return q_1.resolve(this.locationService)
            .then(function (ls) { return ls ? ls.stop() : null; })
            .then(function () { return _this.connection ? _this.connection.controllerExitCode : undefined; });
    };
    HubLookup.prototype.disposeAsync = function () {
        this.disposed = true;
        return q_1.resolve(this.locationService)
            .then(function (ls) { return ls ? ls.dispose() : undefined; });
    };
    HubLookup.prototype.resetLocationService = function () {
        this.locationService = null;
    };
    HubLookup.prototype.throwIfDisposed = function () {
        if (this.disposed) {
            throw new common_1.ServiceHubError(common_1.ErrorKind.ObjectDisposed);
        }
    };
    return HubLookup;
}());
exports.HubLookup = HubLookup;

// SIG // Begin signature block
// SIG // MIIhvAYJKoZIhvcNAQcCoIIhrTCCIakCAQExDzANBglg
// SIG // hkgBZQMEAgEFADB3BgorBgEEAYI3AgEEoGkwZzAyBgor
// SIG // BgEEAYI3AgEeMCQCAQEEEBDgyQbOONQRoqMAEEvTUJAC
// SIG // AQACAQACAQACAQACAQAwMTANBglghkgBZQMEAgEFAAQg
// SIG // T3rXEcL+DcS7RQPZxwruoSpfeHLCDZNeUK9YKrGfG1qg
// SIG // ggtWMIIE3jCCA8agAwIBAgITMwAAAcy1W0IXB2ATEQAA
// SIG // AAABzDANBgkqhkiG9w0BAQsFADB+MQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQgQ29kZSBT
// SIG // aWduaW5nIFBDQSAyMDEwMB4XDTE3MDkxMjE4MDcwNloX
// SIG // DTE4MDkxMjE4MDcwNlowdDELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjEeMBwGA1UEAxMVTWljcm9zb2Z0IENvcnBvcmF0aW9u
// SIG // MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA
// SIG // qZB1moyDMhvorMTex1h+UmWZ52gS7QV0SxKYiTA2P3Ep
// SIG // m3JVmRehPh979B7y6IQgC1C1jrZvtxI21cmxI2KOfkIu
// SIG // 6QywYHZohyxI1JMbfDeVZVCHaafQKW+IZeKZwjAMl2Cw
// SIG // GqeLCmRY3bNRsq5x9ldPKLQX5dZgsCuGeSgrWfQbvlw5
// SIG // 9LTN/Ds5kGkZf2dMSzu6n1lVGgY/7KT75ZCq/p7p4Eru
// SIG // STIPMUnCks35BzTHKwAwEP0QyhnsJGPq3a2Kk/XGmgiU
// SIG // Is9Co0CnbcwsXdXxQceTdcovfituYLlo8qowRTX575Lh
// SIG // dSSJQFyepgmTN94YvAm0bUWaQsK5czBxiwIDAQABo4IB
// SIG // XTCCAVkwHwYDVR0lBBgwFgYKKwYBBAGCNz0GAQYIKwYB
// SIG // BQUHAwMwHQYDVR0OBBYEFFlVxIKPLL3blp451S/Ro/0y
// SIG // ArJ/MDQGA1UdEQQtMCukKTAnMQ0wCwYDVQQLEwRNT1BS
// SIG // MRYwFAYDVQQFEw0yMzA4NjUrMjQyNDA2MB8GA1UdIwQY
// SIG // MBaAFOb8X3u7IgBY5HJOtfQhdCMy5u+sMFYGA1UdHwRP
// SIG // ME0wS6BJoEeGRWh0dHA6Ly9jcmwubWljcm9zb2Z0LmNv
// SIG // bS9wa2kvY3JsL3Byb2R1Y3RzL01pY0NvZFNpZ1BDQV8y
// SIG // MDEwLTA3LTA2LmNybDBaBggrBgEFBQcBAQROMEwwSgYI
// SIG // KwYBBQUHMAKGPmh0dHA6Ly93d3cubWljcm9zb2Z0LmNv
// SIG // bS9wa2kvY2VydHMvTWljQ29kU2lnUENBXzIwMTAtMDct
// SIG // MDYuY3J0MAwGA1UdEwEB/wQCMAAwDQYJKoZIhvcNAQEL
// SIG // BQADggEBANGT7iex7letmRmMI2tjEMw591UYclg78am3
// SIG // zuDKljBRU8yY8BJTGCvoGpikCbUB0SNGoXj3/2XOTlse
// SIG // YC5zfrAJekrn1313An/RxRGJsD8y7Ya38xAJ244R01Yd
// SIG // 9IVXAin+6rhRv0PufFSfH7mdyeAMDE4Y0Dj3j6pw8Cob
// SIG // tPMrYXtwihbvO/faO/afni6lv4RlCSU5sdIgCdq42Q/k
// SIG // J31KgwF17IJ30LmgQyumPqUe9UsDdCWIo2BiJc2NOvaR
// SIG // ChsDvWdCyaOrmpP9elruUdb5hm5M5wBYLEPT+kTIAx6f
// SIG // 78TM4c2wyev6XQjaR4FXESduccHGT35ECK8c7quVDqMw
// SIG // ggZwMIIEWKADAgECAgphDFJMAAAAAAADMA0GCSqGSIb3
// SIG // DQEBCwUAMIGIMQswCQYDVQQGEwJVUzETMBEGA1UECBMK
// SIG // V2FzaGluZ3RvbjEQMA4GA1UEBxMHUmVkbW9uZDEeMBwG
// SIG // A1UEChMVTWljcm9zb2Z0IENvcnBvcmF0aW9uMTIwMAYD
// SIG // VQQDEylNaWNyb3NvZnQgUm9vdCBDZXJ0aWZpY2F0ZSBB
// SIG // dXRob3JpdHkgMjAxMDAeFw0xMDA3MDYyMDQwMTdaFw0y
// SIG // NTA3MDYyMDUwMTdaMH4xCzAJBgNVBAYTAlVTMRMwEQYD
// SIG // VQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRtb25k
// SIG // MR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRpb24x
// SIG // KDAmBgNVBAMTH01pY3Jvc29mdCBDb2RlIFNpZ25pbmcg
// SIG // UENBIDIwMTAwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAw
// SIG // ggEKAoIBAQDpDmRQeWe1xOP9CQBMnpSs91Zo6kTYz8VY
// SIG // T6mldnxtRbrTOZK0pB75+WWC5BfSj/1EnAjoZZPOLFWE
// SIG // v30I4y4rqEErGLeiS25JTGsVB97R0sKJHnGUzbV/S7Sv
// SIG // CNjMiNZrF5Q6k84mP+zm/jSYV9UdXUn2siou1YW7WT/4
// SIG // kLQrg3TKK7M7RuPwRknBF2ZUyRy9HcRVYldy+Ge5JSA0
// SIG // 3l2mpZVeqyiAzdWynuUDtWPTshTIwciKJgpZfwfs/w7t
// SIG // gBI1TBKmvlJb9aba4IsLSHfWhUfVELnG6Krui2otBVxg
// SIG // xrQqW5wjHF9F4xoUHm83yxkzgGqJTaNqZmN4k9Uwz5Uf
// SIG // AgMBAAGjggHjMIIB3zAQBgkrBgEEAYI3FQEEAwIBADAd
// SIG // BgNVHQ4EFgQU5vxfe7siAFjkck619CF0IzLm76wwGQYJ
// SIG // KwYBBAGCNxQCBAweCgBTAHUAYgBDAEEwCwYDVR0PBAQD
// SIG // AgGGMA8GA1UdEwEB/wQFMAMBAf8wHwYDVR0jBBgwFoAU
// SIG // 1fZWy4/oolxiaNE9lJBb186aGMQwVgYDVR0fBE8wTTBL
// SIG // oEmgR4ZFaHR0cDovL2NybC5taWNyb3NvZnQuY29tL3Br
// SIG // aS9jcmwvcHJvZHVjdHMvTWljUm9vQ2VyQXV0XzIwMTAt
// SIG // MDYtMjMuY3JsMFoGCCsGAQUFBwEBBE4wTDBKBggrBgEF
// SIG // BQcwAoY+aHR0cDovL3d3dy5taWNyb3NvZnQuY29tL3Br
// SIG // aS9jZXJ0cy9NaWNSb29DZXJBdXRfMjAxMC0wNi0yMy5j
// SIG // cnQwgZ0GA1UdIASBlTCBkjCBjwYJKwYBBAGCNy4DMIGB
// SIG // MD0GCCsGAQUFBwIBFjFodHRwOi8vd3d3Lm1pY3Jvc29m
// SIG // dC5jb20vUEtJL2RvY3MvQ1BTL2RlZmF1bHQuaHRtMEAG
// SIG // CCsGAQUFBwICMDQeMiAdAEwAZQBnAGEAbABfAFAAbwBs
// SIG // AGkAYwB5AF8AUwB0AGEAdABlAG0AZQBuAHQALiAdMA0G
// SIG // CSqGSIb3DQEBCwUAA4ICAQAadO9XTyl7xBaFeLhQ0yL8
// SIG // CZ2sgpf4NP8qLJeVEuXkv8+/k8jjNKnbgbjcHgC+0jVv
// SIG // r+V/eZV35QLU8evYzU4eG2GiwlojGvCMqGJRRWcI4z88
// SIG // HpP4MIUXyDlAptcOsyEp5aWhaYwik8x0mOehR0PyU6zA
// SIG // DzBpf/7SJSBtb2HT3wfV2XIALGmGdj1R26Y5SMk3YW0H
// SIG // 3VMZy6fWYcK/4oOrD+Brm5XWfShRsIlKUaSabMi3H0oa
// SIG // Dmmp19zBftFJcKq2rbtyR2MX+qbWoqaG7KgQRJtjtrJp
// SIG // iQbHRoZ6GD/oxR0h1Xv5AiMtxUHLvx1MyBbvsZx//CJL
// SIG // SYpuFeOmf3Zb0VN5kYWd1dLbPXM18zyuVLJSR2rAqhOV
// SIG // 0o4R2plnXjKM+zeF0dx1hZyHxlpXhcK/3Q2PjJst67Tu
// SIG // zyfTtV5p+qQWBAGnJGdzz01Ptt4FVpd69+lSTfR3BU+F
// SIG // xtgL8Y7tQgnRDXbjI1Z4IiY2vsqxjG6qHeSF2kczYo+k
// SIG // yZEzX3EeQK+YZcki6EIhJYocLWDZN4lBiSoWD9dhPJRo
// SIG // YFLv1keZoIBA7hWBdz6c4FMYGlAdOJWbHmYzEyc5F3iH
// SIG // Ns5Ow1+y9T1HU7bg5dsLYT0q15IszjdaPkBCMaQfEAjC
// SIG // Vpy/JF1RAp1qedIX09rBlI4HeyVxRKsGaubUxt8jmpZ1
// SIG // xTGCFb4wghW6AgEBMIGVMH4xCzAJBgNVBAYTAlVTMRMw
// SIG // EQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRt
// SIG // b25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRp
// SIG // b24xKDAmBgNVBAMTH01pY3Jvc29mdCBDb2RlIFNpZ25p
// SIG // bmcgUENBIDIwMTACEzMAAAHMtVtCFwdgExEAAAAAAcww
// SIG // DQYJYIZIAWUDBAIBBQCggbAwGQYJKoZIhvcNAQkDMQwG
// SIG // CisGAQQBgjcCAQQwHAYKKwYBBAGCNwIBCzEOMAwGCisG
// SIG // AQQBgjcCARUwLwYJKoZIhvcNAQkEMSIEIF2AUMJaHk+7
// SIG // ipoU326+3/Aly0xP78vaRHXAbctwo7BMMEQGCisGAQQB
// SIG // gjcCAQwxNjA0oBqAGABoAHUAYgBMAG8AbwBrAHUAcAAu
// SIG // AGoAc6EWgBRodHRwOi8vbWljcm9zb2Z0LmNvbTANBgkq
// SIG // hkiG9w0BAQEFAASCAQB/5OfSe9CWpo1PlkL5T+CWSrQK
// SIG // rv7mM2CKUUrREXygYSE1AOtQf+k0xEq0HKg8J7osHz+P
// SIG // 2SMLGCvWgEv7prglXK2Y5LRZ1zgJzMaTbV7DAjar3P7f
// SIG // dv7eysXlxImd6DesrhSdzire434dk67hLX97zQx8IELb
// SIG // Jpbt8TyA77WX8xBP29k89kCuwAFoGixDNWv1mWx7WVpz
// SIG // u5sJtqhOZNbpeGkxRPkPQHJvbOx39uLVw8aqat+Ynlp0
// SIG // VekPjSJt78hkaEouhipUwyBMma4Y6IkE0gegPvQ0OwZy
// SIG // NIV/ycbnxpwtt1CGukccuCmyHSt4JGJAe7CP4VbZs6dl
// SIG // Po0zWgvOoYITRjCCE0IGCisGAQQBgjcDAwExghMyMIIT
// SIG // LgYJKoZIhvcNAQcCoIITHzCCExsCAQMxDzANBglghkgB
// SIG // ZQMEAgEFADCCATwGCyqGSIb3DQEJEAEEoIIBKwSCAScw
// SIG // ggEjAgEBBgorBgEEAYRZCgMBMDEwDQYJYIZIAWUDBAIB
// SIG // BQAEINBWWkFzS9YVCU0aUcptdCEarZcJvAsMJYvhG87n
// SIG // SSJpAgZae1LgekYYEzIwMTgwMjEzMDUxMjM2LjAxM1ow
// SIG // BwIBAYACAfSggbikgbUwgbIxCzAJBgNVBAYTAlVTMRMw
// SIG // EQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRt
// SIG // b25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRp
// SIG // b24xDDAKBgNVBAsTA0FPQzEnMCUGA1UECxMebkNpcGhl
// SIG // ciBEU0UgRVNOOjU3QzgtMkQxNS0xQzhCMSUwIwYDVQQD
// SIG // ExxNaWNyb3NvZnQgVGltZS1TdGFtcCBTZXJ2aWNloIIO
// SIG // yjCCBnEwggRZoAMCAQICCmEJgSoAAAAAAAIwDQYJKoZI
// SIG // hvcNAQELBQAwgYgxCzAJBgNVBAYTAlVTMRMwEQYDVQQI
// SIG // EwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRtb25kMR4w
// SIG // HAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRpb24xMjAw
// SIG // BgNVBAMTKU1pY3Jvc29mdCBSb290IENlcnRpZmljYXRl
// SIG // IEF1dGhvcml0eSAyMDEwMB4XDTEwMDcwMTIxMzY1NVoX
// SIG // DTI1MDcwMTIxNDY1NVowfDELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjEmMCQGA1UEAxMdTWljcm9zb2Z0IFRpbWUtU3RhbXAg
// SIG // UENBIDIwMTAwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAw
// SIG // ggEKAoIBAQCpHQ28dxGKOiDs/BOX9fp/aZRrdFQQ1aUK
// SIG // AIKF++18aEssX8XD5WHCdrc+Zitb8BVTJwQxH0EbGpUd
// SIG // zgkTjnxhMFmxMEQP8WCIhFRDDNdNuDgIs0Ldk6zWczBX
// SIG // JoKjRQ3Q6vVHgc2/JGAyWGBG8lhHhjKEHnRhZ5FfgVSx
// SIG // z5NMksHEpl3RYRNuKMYa+YaAu99h/EbBJx0kZxJyGiGK
// SIG // r0tkiVBisV39dx898Fd1rL2KQk1AUdEPnAY+Z3/1ZsAD
// SIG // lkR+79BL/W7lmsqxqPJ6Kgox8NpOBpG2iAg16HgcsOmZ
// SIG // zTznL0S6p/TcZL2kAcEgCZN4zfy8wMlEXV4WnAEFTyJN
// SIG // AgMBAAGjggHmMIIB4jAQBgkrBgEEAYI3FQEEAwIBADAd
// SIG // BgNVHQ4EFgQU1WM6XIoxkPNDe3xGG8UzaFqFbVUwGQYJ
// SIG // KwYBBAGCNxQCBAweCgBTAHUAYgBDAEEwCwYDVR0PBAQD
// SIG // AgGGMA8GA1UdEwEB/wQFMAMBAf8wHwYDVR0jBBgwFoAU
// SIG // 1fZWy4/oolxiaNE9lJBb186aGMQwVgYDVR0fBE8wTTBL
// SIG // oEmgR4ZFaHR0cDovL2NybC5taWNyb3NvZnQuY29tL3Br
// SIG // aS9jcmwvcHJvZHVjdHMvTWljUm9vQ2VyQXV0XzIwMTAt
// SIG // MDYtMjMuY3JsMFoGCCsGAQUFBwEBBE4wTDBKBggrBgEF
// SIG // BQcwAoY+aHR0cDovL3d3dy5taWNyb3NvZnQuY29tL3Br
// SIG // aS9jZXJ0cy9NaWNSb29DZXJBdXRfMjAxMC0wNi0yMy5j
// SIG // cnQwgaAGA1UdIAEB/wSBlTCBkjCBjwYJKwYBBAGCNy4D
// SIG // MIGBMD0GCCsGAQUFBwIBFjFodHRwOi8vd3d3Lm1pY3Jv
// SIG // c29mdC5jb20vUEtJL2RvY3MvQ1BTL2RlZmF1bHQuaHRt
// SIG // MEAGCCsGAQUFBwICMDQeMiAdAEwAZQBnAGEAbABfAFAA
// SIG // bwBsAGkAYwB5AF8AUwB0AGEAdABlAG0AZQBuAHQALiAd
// SIG // MA0GCSqGSIb3DQEBCwUAA4ICAQAH5ohRDeLG4Jg/gXED
// SIG // PZ2joSFvs+umzPUxvs8F4qn++ldtGTCzwsVmyWrf9efw
// SIG // eL3HqJ4l4/m87WtUVwgrUYJEEvu5U4zM9GASinbMQEBB
// SIG // m9xcF/9c+V4XNZgkVkt070IQyK+/f8Z/8jd9Wj8c8pl5
// SIG // SpFSAK84Dxf1L3mBZdmptWvkx872ynoAb0swRCQiPM/t
// SIG // A6WWj1kpvLb9BOFwnzJKJ/1Vry/+tuWOM7tiX5rbV0Dp
// SIG // 8c6ZZpCM/2pif93FSguRJuI57BlKcWOdeyFtw5yjojz6
// SIG // f32WapB4pm3S4Zz5Hfw42JT0xqUKloakvZ4argRCg7i1
// SIG // gJsiOCC1JeVk7Pf0v35jWSUPei45V3aicaoGig+JFrph
// SIG // pxHLmtgOR5qAxdDNp9DvfYPw4TtxCd9ddJgiCGHasFAe
// SIG // b73x4QDf5zEHpJM692VHeOj4qEir995yfmFrb3epgcun
// SIG // Caw5u+zGy9iCtHLNHfS4hQEegPsbiSpUObJb2sgNVZl6
// SIG // h3M7COaYLeqN4DMuEin1wC9UJyH3yKxO2ii4sanblrKn
// SIG // QqLJzxlBTeCG+SqaoxFmMNO7dDJL32N79ZmKLxvHIa9Z
// SIG // ta7cRDyXUHHXodLFVeNp3lfB0d4wwP3M5k37Db9dT+md
// SIG // Hhk4L7zPWAUu7w2gUDXa7wknHNWzfjUeCLraNtvTX4/e
// SIG // dIhJEjCCBNkwggPBoAMCAQICEzMAAACqt6mI/+pXwwoA
// SIG // AAAAAKowDQYJKoZIhvcNAQELBQAwfDELMAkGA1UEBhMC
// SIG // VVMxEzARBgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcT
// SIG // B1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jw
// SIG // b3JhdGlvbjEmMCQGA1UEAxMdTWljcm9zb2Z0IFRpbWUt
// SIG // U3RhbXAgUENBIDIwMTAwHhcNMTYwOTA3MTc1NjUzWhcN
// SIG // MTgwOTA3MTc1NjUzWjCBsjELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjEMMAoGA1UECxMDQU9DMScwJQYDVQQLEx5uQ2lwaGVy
// SIG // IERTRSBFU046NTdDOC0yRDE1LTFDOEIxJTAjBgNVBAMT
// SIG // HE1pY3Jvc29mdCBUaW1lLVN0YW1wIFNlcnZpY2UwggEi
// SIG // MA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQCe2H97
// SIG // V3j6BfqjmreQr5q51bZ0dpHLuM67Ks4rKe9ftGt1mHUo
// SIG // gK2Yj7mq8J3StptHZJjGpVmqkXj5Hot59yR0Ruok/mkE
// SIG // rr7rf1NTE1nVD/z7zd32B0GbGzc32Vx2md2ux8Pb061o
// SIG // wlCcdA5Edy4unW/BHe8sxzt0rtuXLY+BxUcFQMBKN0iT
// SIG // vAOaS/CNBMtDcPvoWLGdrFDnubWE8cebB+z3mmnbr0h/
// SIG // rFkxhmIWcJVe8cvMOkfz6j8CjBC7vSOMqQNzTw64DFzx
// SIG // L6p/+mSbjC6YFMBGzyZPPAYjxk8it5IPGgIZ/JwpzYTe
// SIG // ga/w/a2YrKWIg4aVVIa0m3FAxi83AgMBAAGjggEbMIIB
// SIG // FzAdBgNVHQ4EFgQUCLhyHRUk9ZMC4Z/SHEzTq8xZybAw
// SIG // HwYDVR0jBBgwFoAU1WM6XIoxkPNDe3xGG8UzaFqFbVUw
// SIG // VgYDVR0fBE8wTTBLoEmgR4ZFaHR0cDovL2NybC5taWNy
// SIG // b3NvZnQuY29tL3BraS9jcmwvcHJvZHVjdHMvTWljVGlt
// SIG // U3RhUENBXzIwMTAtMDctMDEuY3JsMFoGCCsGAQUFBwEB
// SIG // BE4wTDBKBggrBgEFBQcwAoY+aHR0cDovL3d3dy5taWNy
// SIG // b3NvZnQuY29tL3BraS9jZXJ0cy9NaWNUaW1TdGFQQ0Ff
// SIG // MjAxMC0wNy0wMS5jcnQwDAYDVR0TAQH/BAIwADATBgNV
// SIG // HSUEDDAKBggrBgEFBQcDCDANBgkqhkiG9w0BAQsFAAOC
// SIG // AQEAYI8O7gfOGuF9n3nGeA6dzkykAZ1qZ0eXERuKcsrk
// SIG // wXznLeOPkWe86HR6P8rpRiQAP6HO8H5P0vQaffR2OB0U
// SIG // Nh2l2YlvysVTFO8TLCACldEKecXS7m08P5FG6blS3t9c
// SIG // 4pykTVFqHcLpk01GchYm+YT/k3fd6AM9VPzCBKBfj4e9
// SIG // VXSa6WSssOQaylw7IB8LVVgIsMPLp7xZLE1Cke1bszAu
// SIG // kqeTjk6ADK6peTHsUpF8lRCvf8HOI9sPmcxqw8T0LB91
// SIG // ZIIsoNgOB/eaDmWoXJWBnH5Y7nnzkSGt280sv7WIcv4G
// SIG // G51fdg92MoiuUjxtOS7MBk4kSS0vqWYA1vxoJqGCA3Qw
// SIG // ggJcAgEBMIHioYG4pIG1MIGyMQswCQYDVQQGEwJVUzET
// SIG // MBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMHUmVk
// SIG // bW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBvcmF0
// SIG // aW9uMQwwCgYDVQQLEwNBT0MxJzAlBgNVBAsTHm5DaXBo
// SIG // ZXIgRFNFIEVTTjo1N0M4LTJEMTUtMUM4QjElMCMGA1UE
// SIG // AxMcTWljcm9zb2Z0IFRpbWUtU3RhbXAgU2VydmljZaIl
// SIG // CgEBMAkGBSsOAwIaBQADFQCcnMVrmMynwa1HIOzTRPzO
// SIG // bdDqA6CBwTCBvqSBuzCBuDELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjEMMAoGA1UECxMDQU9DMScwJQYDVQQLEx5uQ2lwaGVy
// SIG // IE5UUyBFU046MjY2NS00QzNGLUM1REUxKzApBgNVBAMT
// SIG // Ik1pY3Jvc29mdCBUaW1lIFNvdXJjZSBNYXN0ZXIgQ2xv
// SIG // Y2swDQYJKoZIhvcNAQEFBQACBQDeLIT0MCIYDzIwMTgw
// SIG // MjEyMjEyNjEyWhgPMjAxODAyMTMyMTI2MTJaMHQwOgYK
// SIG // KwYBBAGEWQoEATEsMCowCgIFAN4shPQCAQAwBwIBAAIC
// SIG // CBIwBwIBAAICGSwwCgIFAN4t1nQCAQAwNgYKKwYBBAGE
// SIG // WQoEAjEoMCYwDAYKKwYBBAGEWQoDAaAKMAgCAQACAxbj
// SIG // YKEKMAgCAQACAx6EgDANBgkqhkiG9w0BAQUFAAOCAQEA
// SIG // WfQbhpvJqZfba9rMD88gXlvOBxPf/yqTEMd+QOmciF65
// SIG // 67hTOBqG/T14hDl0CoO3zDCAW2D/iv4n0QsDEE7IvkRH
// SIG // j8psD54YilymWnWi7I5nrnceuStlesPaBUdagPRfazQ0
// SIG // jTQk7FTvWOOEKrkeYwYFUqpPT0vm3rUJ4j2gMb7bUW9y
// SIG // kxQQj8yKo6zkdspsgQenaWfEPhwZQqenMPD7DK16blaH
// SIG // mpvUjWpo7Vk28BHcZ0Aahi1fj+cAoyOeEGqtKsX6VjOI
// SIG // xVhcE0SOqi7CnbbgBUSe5GajDo5+tU0Er+L6CUzuDMMZ
// SIG // JRJYmBcvMWXfvZQtfmsu0SLuy9Dc5ZG7HjGCAvUwggLx
// SIG // AgEBMIGTMHwxCzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpX
// SIG // YXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRtb25kMR4wHAYD
// SIG // VQQKExVNaWNyb3NvZnQgQ29ycG9yYXRpb24xJjAkBgNV
// SIG // BAMTHU1pY3Jvc29mdCBUaW1lLVN0YW1wIFBDQSAyMDEw
// SIG // AhMzAAAAqrepiP/qV8MKAAAAAACqMA0GCWCGSAFlAwQC
// SIG // AQUAoIIBMjAaBgkqhkiG9w0BCQMxDQYLKoZIhvcNAQkQ
// SIG // AQQwLwYJKoZIhvcNAQkEMSIEIBBebNmF/A2voM25MVWV
// SIG // xwzyKvAfulPwKHSwrUjHnmmyMIHiBgsqhkiG9w0BCRAC
// SIG // DDGB0jCBzzCBzDCBsQQUnJzFa5jMp8GtRyDs00T8zm3Q
// SIG // 6gMwgZgwgYCkfjB8MQswCQYDVQQGEwJVUzETMBEGA1UE
// SIG // CBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMHUmVkbW9uZDEe
// SIG // MBwGA1UEChMVTWljcm9zb2Z0IENvcnBvcmF0aW9uMSYw
// SIG // JAYDVQQDEx1NaWNyb3NvZnQgVGltZS1TdGFtcCBQQ0Eg
// SIG // MjAxMAITMwAAAKq3qYj/6lfDCgAAAAAAqjAWBBTnsK3Y
// SIG // w8cPb2SLZM1M1OgNJrP/wTANBgkqhkiG9w0BAQsFAASC
// SIG // AQAERVDfjd+WeoVb4KJLQgN+cDIvJskkgL1b6IXdc+rY
// SIG // MlqpFdFdszY+yzRDG/Y802BSn9QaKlVtCkMPPxhs7GlG
// SIG // RB4ImIO9rgZ7gt4pVWqt+EgHHXI10u/n1zGl03Iei+aH
// SIG // CZ77D4SL7vZNyOvpL41O7QF52QPq69DIoTe97hS2odPc
// SIG // zChlihVaz8qMp1efwhe6oUMH0Dp4C3yFpfM7m/YJfIJR
// SIG // TRYiSwZWxZvu5TwVSPXCn5n/5W+IGZFYDuzYmsBasNy9
// SIG // m6jf6pRdrxV9pbZfKxzUA+10W7RhOnnjMOrYUEPoCdRb
// SIG // I0ZFG/ul23Rn6DkFTJa54TdTRII2/TTI0bsC
// SIG // End signature block
