// vs-telemetry/internal/actions/action.js
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/
"use strict";
(function (ActionPriority) {
    ActionPriority[ActionPriority["ManifestExcludePriority"] = 0] = "ManifestExcludePriority";
    ActionPriority[ActionPriority["CustomOptOutPriority"] = 1] = "CustomOptOutPriority";
    ActionPriority[ActionPriority["ManifestOptOutPriority"] = 2] = "ManifestOptOutPriority";
    ActionPriority[ActionPriority["ManifestPiiPriority"] = 3] = "ManifestPiiPriority";
    ActionPriority[ActionPriority["MetricPriority"] = 50] = "MetricPriority";
    ActionPriority[ActionPriority["SettingPriority"] = 60] = "SettingPriority";
    ActionPriority[ActionPriority["PiiPriority"] = 100] = "PiiPriority";
    ActionPriority[ActionPriority["ThrottlingPriority"] = 1000] = "ThrottlingPriority";
    ActionPriority[ActionPriority["ManifestRoutePriority"] = 10000] = "ManifestRoutePriority";
    ActionPriority[ActionPriority["Router"] = 1000000000] = "Router";
})(exports.ActionPriority || (exports.ActionPriority = {}));
var ActionPriority = exports.ActionPriority;
