// vs-telemetry/internal/telemetry-manifest/telemetry-manifest-action-opt-out-properties.js
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var index_1 = require("../../index");
var telemetry_manifest_action_opt_out_events_1 = require("./telemetry-manifest-action-opt-out-events");
var telemetry_manifest_1 = require("./telemetry-manifest");
var stringjs = require("string");
var ManifestActionOptOutProperties = (function (_super) {
    __extends(ManifestActionOptOutProperties, _super);
    function ManifestActionOptOutProperties() {
        _super.apply(this, arguments);
        this._properties = [];
    }
    Object.defineProperty(ManifestActionOptOutProperties.prototype, "properties", {
        get: function () {
            return this._properties;
        },
        set: function (value) {
            if (value == null) {
                throw new index_1.ArgumentException("properties cannot be null");
            }
            this._properties = value;
        },
        enumerable: true,
        configurable: true
    });
    ManifestActionOptOutProperties.prototype.validate = function () {
        if (this.properties.length === 0) {
            throw new telemetry_manifest_1.TelemetryManifestValidationException("properties must have at least one property");
        }
        if (this.properties.some(function (propertyName) { return stringjs(propertyName).isEmpty(); })) {
            throw new telemetry_manifest_1.TelemetryManifestValidationException("properties must not contain empty/null property name");
        }
    };
    ManifestActionOptOutProperties.prototype.executeOptOutAction = function (context) {
        var _this = this;
        this.properties.forEach(function (propertyName) {
            _this.processPropertyName(propertyName, context);
        });
    };
    return ManifestActionOptOutProperties;
}(telemetry_manifest_action_opt_out_events_1.ManifestActionOptOutBase));
exports.ManifestActionOptOutProperties = ManifestActionOptOutProperties;
var ManifestActionOptOutExcludeProperties = (function (_super) {
    __extends(ManifestActionOptOutExcludeProperties, _super);
    function ManifestActionOptOutExcludeProperties() {
        _super.apply(this, arguments);
    }
    ManifestActionOptOutExcludeProperties.prototype.processPropertyName = function (propertyName, context) {
        context.excludePropertyFromEvent(propertyName);
    };
    return ManifestActionOptOutExcludeProperties;
}(ManifestActionOptOutProperties));
exports.ManifestActionOptOutExcludeProperties = ManifestActionOptOutExcludeProperties;
var ManifestActionOptOutIncludeProperties = (function (_super) {
    __extends(ManifestActionOptOutIncludeProperties, _super);
    function ManifestActionOptOutIncludeProperties() {
        _super.apply(this, arguments);
    }
    ManifestActionOptOutIncludeProperties.prototype.processPropertyName = function (propertyName, context) {
        context.includePropertyToEvent(propertyName);
    };
    return ManifestActionOptOutIncludeProperties;
}(ManifestActionOptOutProperties));
exports.ManifestActionOptOutIncludeProperties = ManifestActionOptOutIncludeProperties;
