// vs-telemetry/internal/telemetry-manifest/telemetry-manifest-action-opt-out-events.js
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var action_1 = require("../actions/action");
var ManifestActionOptOutBase = (function () {
    function ManifestActionOptOutBase() {
        this.priority = action_1.ActionPriority.ManifestOptOutPriority;
    }
    ManifestActionOptOutBase.prototype.validate = function () {
        return null;
    };
    ManifestActionOptOutBase.prototype.run = function (context) {
        var _this = this;
        return context.telemetrySession.isOptedIn().then(function (optedIn) {
            if (!optedIn) {
                _this.executeOptOutAction(context);
            }
        });
    };
    return ManifestActionOptOutBase;
}());
exports.ManifestActionOptOutBase = ManifestActionOptOutBase;
var ManifestActionOptOutExcludeEvents = (function (_super) {
    __extends(ManifestActionOptOutExcludeEvents, _super);
    function ManifestActionOptOutExcludeEvents() {
        _super.apply(this, arguments);
    }
    ManifestActionOptOutExcludeEvents.prototype.executeOptOutAction = function (context) {
        context.isDropped = true;
    };
    return ManifestActionOptOutExcludeEvents;
}(ManifestActionOptOutBase));
exports.ManifestActionOptOutExcludeEvents = ManifestActionOptOutExcludeEvents;
var ManifestActionOptOutIncludeEvents = (function (_super) {
    __extends(ManifestActionOptOutIncludeEvents, _super);
    function ManifestActionOptOutIncludeEvents() {
        _super.apply(this, arguments);
    }
    ManifestActionOptOutIncludeEvents.prototype.executeOptOutAction = function (context) {
        context.isDropped = false;
    };
    return ManifestActionOptOutIncludeEvents;
}(ManifestActionOptOutBase));
exports.ManifestActionOptOutIncludeEvents = ManifestActionOptOutIncludeEvents;
