// vs-telemetry/internal/registry-tools.js
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/
"use strict";
var registry_tools_native_1 = require("./registry-tools-native");
var registry_tools_native_2 = require("./registry-tools-native");
exports.RegistryArch = registry_tools_native_2.Arch;
var RegistryTools = (function () {
    function RegistryTools(arch, nativeRegistryTools) {
        this._arch = registry_tools_native_1.Arch.Default;
        if (arch != null) {
            this._arch = arch;
        }
        if (nativeRegistryTools == null) {
            nativeRegistryTools = new registry_tools_native_1.NativeRegistryTools(this._arch);
        }
        this._nativeRegistryTools = nativeRegistryTools;
    }
    /**
     * Gets a string from the HKCU hive in the registry.
     * @param key key in the registry
     * @param name name of the registry string
     * @return Promise on completion
     */
    /* istanbul ignore next */
    RegistryTools.prototype.getStringFromHKCU = function (key, name) {
        var _this = this;
        return this._executeRegistryCall(function () { return _this._nativeRegistryTools.getStringFromHKCU(key, name); });
    };
    /**
     * Gets a string from the HKLM hive in the registry.
     * @param key key in the registry
     * @param name name of the registry string
     * @return Promise on completion
     */
    /* istanbul ignore next */
    RegistryTools.prototype.getStringFromHKLM = function (key, name) {
        var _this = this;
        return this._executeRegistryCall(function () { return _this._nativeRegistryTools.getStringFromHKLM(key, name); });
    };
    /**
     * Gets a number from the HKCU hive in the registry.
     * @param key key in the registry
     * @param name name of the registry string
     * @return Promise on completion
     */
    RegistryTools.prototype.getNumberFromHKCU = function (key, name) {
        var _this = this;
        return this._executeRegistryCall(function () { return _this._nativeRegistryTools.getNumberFromHKCU(key, name); });
    };
    /**
     * Gets a number from the HKLM hive in the registry.
     * @param key key in the registry
     * @param name name of the registry string
     * @return Promise on completion
     */
    RegistryTools.prototype.getNumberFromHKLM = function (key, name) {
        var _this = this;
        return this._executeRegistryCall(function () { return _this._nativeRegistryTools.getNumberFromHKLM(key, name); });
    };
    /**
     * Sets a string in the HKCU hive in the registry.
     * @param key key in the registry
     * @param name name of the registry string to setEnabled
     * @param value value of the item to set
     * @return Promise on completion
     */
    /* istanbul ignore next */
    RegistryTools.prototype.setStringInHKCU = function (key, name, value) {
        var _this = this;
        return this._executeRegistryCall(function () { return _this._nativeRegistryTools.setStringInHKCU(key, name, value); });
    };
    /**
     * Sets a string in the HKLM hive in the registry.
     * @param key key in the registry
     * @param name name of the registry string to setEnabled
     * @param value value of the item to set
     * @return Promise on completion
     */
    /* istanbul ignore next */
    RegistryTools.prototype.setStringInHKLM = function (key, name, value) {
        var _this = this;
        return this._executeRegistryCall(function () { return _this._nativeRegistryTools.setStringInHKLM(key, name, value); });
    };
    /**
     * Gets a map of key value pairs from the HKCU hive in the registry.
     * @param key key under which all value will be retrieved
     * @return Promise on completion with the map
     */
    /* istanbul ignore next */
    RegistryTools.prototype.getValuesFromHKCU = function (key) {
        var _this = this;
        return this._executeRegistryCall(function () { return _this._nativeRegistryTools.getValuesFromHKCU(key); });
    };
    /**
     * Gets a map of key value pairs from the HKLM hive in the registry.
     * @param key key under which all value will be retrieved
     * @return Promise on completion with the map
     */
    /* istanbul ignore next */
    RegistryTools.prototype.getValuesFromHKLM = function (key) {
        var _this = this;
        return this._executeRegistryCall(function () { return _this._nativeRegistryTools.getValuesFromHKLM(key); });
    };
    /**
     * Removes the specified value from the HKCU hive in the registry.
     * @param key key in the registry
     * @param name name of the value to remove
     * @return Promise on completion
     */
    /* istanbul ignore next */
    RegistryTools.prototype.removeValueFromHKCU = function (key, name) {
        var _this = this;
        return this._executeRegistryCall(function () { return _this._nativeRegistryTools.removeValueFromHKCU(key, name); });
    };
    /**
     * Removes the specified value from the HKCU hive in the registry.
     * @param key key in the registry
     * @param name name of the value to remove
     * @return Promise on completion
     */
    /* istanbul ignore next */
    RegistryTools.prototype.removeValueFromHKLM = function (key, name) {
        var _this = this;
        return this._executeRegistryCall(function () { return _this._nativeRegistryTools.removeValueFromHKLM(key, name); });
    };
    RegistryTools.prototype._executeRegistryCall = function (nativeWinRegCall) {
        return this._wrapNativeRegistryCall(function () { return nativeWinRegCall(); });
    };
    RegistryTools.prototype._wrapNativeRegistryCall = function (action) {
        try {
            return Promise.resolve(action());
        }
        catch (err) {
            return Promise.reject(err);
        }
    };
    return RegistryTools;
}());
exports.RegistryTools = RegistryTools;
